### PR TITLE
Soap request with complexType throws AssertError

### PR DESCRIPTION
**Pull request by phanikiranwtc**
Below is the sample of the one operation from my wsdl

```
<wsdl:operation name="queryData">
<wsdl:input name="queryDataRequest" message="tns:queryDataRequest"/>
<wsdl:output name="queryDataResponse" message="tns:queryDataResponse"/>
</wsdl:operation>

<wsdl:message name="queryDataRequest">
<wsdl:part name="ModelCRUDRequest" element="tns:ModelCRUDRequest"/>
</wsdl:message>


<xsd:complexType name="ModelCRUDRequest">
<xsd:sequence>
<xsd:element maxOccurs="1" minOccurs="1" name="ModelCRUD" type="tns:ModelCRUD"/>
<xsd:element maxOccurs="1" minOccurs="1" name="ADLoginRequest" type="tns:ADLoginRequest"/>
</xsd:sequence>
</xsd:complexType>


<xsd:complexType name="ModelCRUD">
<xsd:sequence>
<xsd:element name="serviceType" type="xsd:string"/>
<xsd:element name="TableName" type="xsd:string"/>
<xsd:element name="RecordID" type="xsd:int"/>
<xsd:element name="Filter" type="xsd:string"/>
<xsd:element name="Action">
<xsd:simpleType>
<xsd:restriction base="xsd:string">
<xsd:enumeration value="Create"/>
<xsd:enumeration value="Read"/>
<xsd:enumeration value="Update"/>
<xsd:enumeration value="Delete"/>
</xsd:restriction>
</xsd:simpleType>
</xsd:element>
<xsd:element maxOccurs="1" minOccurs="0" name="DataRow" type="tns:DataRow"/>
</xsd:sequence>
</xsd:complexType>

<xsd:complexType name="ADLoginRequest">
<xsd:sequence>
<xsd:element name="user" type="xsd:string"/>
<xsd:element name="pass" type="xsd:string"/>
<xsd:element name="lang" type="xsd:string"/>
<xsd:element name="ClientID" type="xsd:int"/>
<xsd:element name="RoleID" type="xsd:int"/>
<xsd:element name="OrgID" type="xsd:int"/>
<xsd:element name="WarehouseID" type="xsd:int"/>
<xsd:element name="stage" type="xsd:int"/>
</xsd:sequence>
</xsd:complexType>
```

for the above operation the input envelop should be generated as below

`<tns:createData><tns:ModelCRUDRequest><tns:ModelCRUD><tns:serviceType>CreateInterest</tns:serviceType><tns:TableName>WTC_EmployeeInterest</tns:TableName><tns:RecordID>0</tns:RecordID><tns:Filter></tns:Filter><tns:Action>Create</tns:Action><tns:DataRow><tns:field column="C_BPartner_ID"><tns:val>1001047</tns:val></tns:field><tns:field column="Interest"><tns:val>Cooking</tns:val></tns:field></tns:DataRow></tns:ModelCRUD><tns:ADLoginRequest><tns:user>WTCAdmin</tns:user><tns:pass>WTCAdmin</tns:pass><tns:lang>192</tns:lang><tns:ClientID>11</tns:ClientID><tns:RoleID>102</tns:RoleID><tns:OrgID>1000001</tns:OrgID><tns:WarehouseID>1000000</tns:WarehouseID><tns:stage>0</tns:stage></tns:ADLoginRequest></tns:ModelCRUDRequest></tns:createData>`

where as the current version throws AssertError.

Which is fixed in this commit.

### CODE REVIEW

**Comment by phanikiranwtc** [comment]
Hello Team, please ignore the last commit which has myeaglesample.js
ill try to revert the commit.

**Comment by jsdevel** [comment]
Can you please add a test?

**Comment by phanikiranwtc** [comment]
Hello Joseph,
i understand the importance of test cases
i tried to write a test case for the same, but could not do that actually.
i have tested most of the uses cases manually none of them fails.

highly appreciate if you can help me in this case.

**Comment by jsdevel** [comment]
Seems a request/response sample would be good.  Checkout the folder under `test/`.

**Comment by phanikiranwtc** [comment]
Hello Joseph,
Seems like need your help.
i have added the complete wsdl file in soap.wsdl file just changed the port address to http://localhost:1509

request and response i have copied from the input and output generated by the code with the mychanges.

Can you please let me what am missing here the test cases are failed.

Thanks,
Phani

File: test/request-response-samples/complexType_request/request.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (herom) [inline test/request-response-samples/complexType_request/request.json]: This is a malformed JSON file - unlike plain JavaScript objects, in JSON all Strings have to be double quoted -->
for example:
```json
{
  "ModelCRUDRequest": {
    "ModelCRUD": {
      "serviceType": "CreateInterest"
    }
  }
}

File: test/request-response-samples/complexType_request/response.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (herom) [inline test/request-response-samples/complexType_request/response.json]: This JSON file is also malformed.

File: test/request-response-samples/complexType_request/soap.wsdl [hunk 3]
```diff
@@ -0,0 +1,634 @@
+<wsdl:definitions xmlns:soapenc12="http://www.w3.org/2003/05/soap-encoding" xmlns:tns="http://3e.pl/ADInterface" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap11="http://schemas.xmlsoap.org/soap/envelope/" xmlns:wsdlsoap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:soapenc11="http://schemas.xmlsoap.org/soap/encoding/" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope" targetNamespace="http://3e.pl/ADInterface">
```
Comment (herom) [inline test/request-response-samples/complexType_request/soap.wsdl]: Please shorten the `wsdl` file - at best it should only contain the things you need for testing. This helps us to keep the library size rather small :+1:

File: lib/wsdl.js [hunk 4]
```diff
@@ -1528,31 +1528,30 @@ WSDL.prototype.objectToDocumentXML = function(name, params, nsPrefix, nsURI, typ
...
+
+   nsPrefix = nsPrefix || findPrefix(defs.xmlns, nsURI);
+
+   nsURI = nsURI || defs.xmlns[nsPrefix];
+   nsPrefix = nsPrefix === TNS_PREFIX ? '' : (nsPrefix + ':');
+
+   parts.push(['<', nsPrefix, name, '>'].join(''));
+
+   for (var key in params) {
+     if (!params.hasOwnProperty(key)) continue;
```
Comment (herom) [inline lib/wsdl.js]: Please don't make "one liners" for `if` constructs but use curly braces instead

File: lib/wsdl.js [hunk 5]
```diff
@@ -1528,31 +1528,30 @@ WSDL.prototype.objectToDocumentXML = function(name, params, nsPrefix, nsURI, typ
...
+     if (key !== nsAttrName) {
+       var value = params[key];
+       parts.push(['<', (isParts ? '' : nsPrefix), key, '>'].join(''));
```
Comment (herom) [inline lib/wsdl.js]: Instead of having to check `isParts ? '' : nsPrefix` two times, wouldn't it be also sufficient to create athe key right away like 
```javascript
var value = params[key];
var prefixedKey = (isParts ? '' : nsPrefix) + key;
parts.push(['<', prefixedKey, '>].join('')); 
parts.push((typeof value === 'object') ? this.objectToXML(value, key, nsPrefix, nsURI) : xmlEscape(value));
parts.push(['</', prefixedKey, '>].join(''));
```

**Comment by herom** [comment]
The tests over at Travis are failing because the request/response samples you're provided are not named after the needed schema. Please rename them to reflect the used naming scheme which expects the folder to be named like `{prefix, like 'Dummy', etc.}__{name}_{your}_{test}`.

As you can see the tests are failing because of the test loading logic of the `request-response-sample-tests.js` cannot operate with a name which violates this naming scheme.

You can start tests at your local machine with the command `npm test`.

**Comment by phanikiranwtc** [comment]
Thanks Herom for your input.
i have fixed the request/response samples. now all the tests are passed.

File: package.json [hunk 6]
```diff
@@ -47,13 +46,16 @@
     "diff": "^2.2.1",
     "doctoc": "^1.0.0",
     "duplexer": "~0.1.1",
+    "generate-function": "^2.0.0",
```
Comment (herom) [inline package.json]: why was this library included? I don't see any `require('generate-function')`s within the code?

File: package.json [hunk 7]
```diff
@@ -32,7 +32,6 @@
   },
   "scripts": {
     "toc": "./node_modules/.bin/doctoc Readme.md --github --maxlevel 3",
-    "pretest": "jshint index.js lib test",
```
Comment (herom) [inline package.json]: why was this removed?

File: package.json [hunk 8]
```diff
@@ -47,13 +46,16 @@
     "diff": "^2.2.1",
     "doctoc": "^1.0.0",
     "duplexer": "~0.1.1",
+    "generate-function": "^2.0.0",
     "glob": "~3.2.8",
     "istanbul": "^0.4.1",
-    "jshint": "2.3.0",
-    "mocha": "~1.17.0",
+    "jshint": "^2.9.2",
+    "jsonpointer": "^3.0.0",
```
Comment (herom) [inline package.json]: same as with `generate-function`...

File: package.json [hunk 9]
```diff
@@ -47,13 +46,16 @@
...
+    "mocha": "^2.4.5",
     "readable-stream": "~2.0.2",
     "semver": "~5.0.3",
     "should": "~3.3.0",
-    "timekeeper": "~0.0.4"
+    "timekeeper": "~0.0.4",
+    "xtend": "^4.0.1"
```
Comment (herom) [inline package.json]: same as with `generate-function`

**Comment by phanikiranwtc** [comment]
Hi Herom,
all these changes were made in my local to run the test. when i run npm test. those packages were missing.

unfortunately i have checked in package.json.. ill check in the old package.json right now.

Thanks,
Phani

**Comment by herom** [comment]
Thanks a ton @phanikiranwtc :+1: 

Could you please squash your commits and rebase onto master? see our [Guidelines on Submitting a Pull Request](https://github.com/vpulim/node-soap/blob/master/CONTRIBUTING.md#submitting-a-pull-request)

**Comment by phanikiranwtc** [comment]
Hi Heinz,
Thanks for the support.
I have tried to squash all the commits.
please let me know if this is not meet the need. ill fork the repository fresh and include my changes and push them in single commit and create a pull request.

Thanks,
Phani

**Comment by phanikiranwtc** [comment]
Hi Heinz,
assuming your not comfortable with the commit squash in the pull request.
ill close this pull request and create a new one with single commit in PR with the all the required details.



Thanks,
Phani

**Comment by herom** [comment]
hey @phanikiranwtc - sorry for the lack of response. I'm tight on schedule and mostly cut off from internet in my spare time so I didn't have any time to look at it closely. please don't close this and open another PR but instead take a look at the linked guides in our [Guidelines for Submitting a Pull Request](https://github.com/vpulim/node-soap/blob/master/CONTRIBUTING.md#submitting-a-pull-request) and follow the steps in the great github help pages in order to get your commits squashed.

**Comment by phanikiranwtc** [comment]
Hello Heinz,
I have rebased the commit and pushed the code to master.

Please review.

Thanks,
Phani

**Comment by jsdevel** [comment]
closing due to inactivity
